### PR TITLE
3335 allow alternative character encoding length in searchparam

### LINKED ISSUE #3335: Update phoneticencoders to allow defining custom max code length (for supported phoneticencoders)
Update hapi-fhir to allow defining custom max code length for any supported string encoders. This should allow better fuzzy matching where supported.

### PR DESCRIPTION
**Pull request by TipzCM**
closes #3335 

Added support to allow specifying max code length for phonetic encoders (for supported encoders only - metaphone and double_metaphone).
Added change log

### CODE REVIEW

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderEnum.java [hunk 1]
```diff
@@ -44,10 +44,23 @@
 
 	private final IPhoneticEncoder myPhoneticEncoder;
 
+	/**
+	 * Do not construct this enum via constructor.
+	 *
+	 * Use PhoneticEncoderWrapper instead.
```
Comment (tadgh) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderEnum.java:50]: use {@link PhoneticEncoderWrapper} notation here for easy tracing.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	/**
+	 * Creates the phonetic encoder wrapper from
+	 * an input string.
+	 *
+	 * String must be in the format of...
+	 *
+	 * PhoneticEncoderEnum(MAX_LENGTH)
+	 */
```
Comment (tadgh) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java:36]: If you're going to write javadocs, may as well use the standard format: 

```
	/**
	 * Subclasses should override this method to supply the set of compartments that
	 * the user making the request should actually have access to.
	 *
	 * <p>
	 * Typically this is done by examining <code>theRequestDetails</code> to find
	 * out who the current user is and then building a list of Strings.
	 * </p>
	 *
	 * @param theRequestDetails The individual request currently being applied
	 *
	 * @return The list of allowed compartments and instances that should be used
	 * for search narrowing. If this method returns <code>null</code>, no narrowing will
	 * be performed
	 */
```

```

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+				} catch (NumberFormatException ex) {
+					encoderType = null;
+					ourLog.error("Invalid encoder max character length: " + num);
+				}
+			}
+			// else - parse error
+		}
+		else {
+			encoderType = theString;
+		}
```
Comment (tadgh) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java:59]: suggest refactoring the number-finder into another method, and call it instead of doing the finding in-line

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3335-added-configuration-to-specify-max-code-lengths.yml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+type: add
+issue: 3335
+smile: 3603
```
Comment (tadgh) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3335-added-configuration-to-specify-max-code-lengths.yml:4]: `smile` is not a key that goes into changelogs. If anything, you are looking for `jira: SMILE-1234` for the jira link.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3335-added-configuration-to-specify-max-code-lengths.yml [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+title: "Added ability to specify max code lengths for supported
+      Phonetic Encoders (Metaphone, Double_Metaphone)."
```
Comment (tadgh) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3335-added-configuration-to-specify-max-code-lengths.yml:6]: indicate _how_ to do this somewhere. In [documentation](https://smilecdr.com/docs/fhir_repository/search_parameter_phonetic.html), and also this changelog, for those who only use hapi-fhir

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchCustomSearchParamTest.java [hunk 6]
```diff
@@ -197,6 +201,60 @@ public void testNumberSearchParam() {
...
+		RuntimeSearchParam parammodified = mySearchParamRegistry.getActiveSearchParam("Patient",
+			"fuzzymodified");
+
+		// verify the encoders are different!
+		Assertions.assertNotEquals(paramdefault, parammodified);
+		String encodedDefault = paramdefault.encode(testString);
+		String encodedMod = parammodified.encode(testString);
+		Assertions.assertEquals(modifiedLength, encodedMod.length());
+		Assertions.assertNotEquals(encodedDefault.length(), encodedMod.length());
+	}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchCustomSearchParamTest.java:242]: very nice test!

Generally for asserts, we import them individually, so you just do `assertNotEquals` without having to qualify with `Assertions.`

File: hapi-fhir-base/src/test/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapperTest.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+		// verify
+		Assertions.assertNull(retEnum);
+		ArgumentCaptor<ILoggingEvent> captor = ArgumentCaptor.forClass(ILoggingEvent.class);
+		Mockito.verify(myListAppender)
+			.doAppend(captor.capture());
+		Assertions.assertEquals(1, captor.getAllValues().size());
+		ILoggingEvent event = captor.getValue();
+		Assertions.assertEquals("Invalid phonetic param string " + theString,
+			event.getMessage());
+	}
```
Comment (tadgh) [inline hapi-fhir-base/src/test/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapperTest.java:88]: I would add tests for more weird edge cases. e.g.

`METAPHONE()`
`METAPHONE(-1)`
`METAPHONE(INT_MAX)` :P 
`METAPHONE(()`

**Review by tadgh — CHANGES_REQUESTED** [review]

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+			return null;
+		}
+	}
+
+	private static IPhoneticEncoder getEncoderFromString(String theName, int theMax) {
+		IPhoneticEncoder encoder = null;
+		PhoneticEncoderEnum enumVal = EnumUtils.getEnum(PhoneticEncoderEnum.class, theName);
+
+		if (enumVal != null) {
+			switch (enumVal) {
```
Comment (fil512) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java:76]: Nitpick: We should use polymorphism for this. Not a switch statement

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+import org.apache.commons.codec.language.MatchRatingApproachEncoder;
+import org.apache.commons.codec.language.Metaphone;
+import org.apache.commons.codec.language.Nysiis;
+import org.apache.commons.codec.language.RefinedSoundex;
+import org.apache.commons.codec.language.Soundex;
+import org.apache.commons.lang3.EnumUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PhoneticEncoderWrapper {
```
Comment (fil512) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java:16]: Rather than calling it a wrapper (which I feel is a specific design pattern that doesn't really apply here) can we instead call it a ParameterizedPhoneticEncoder.  And can it interoperate with a regular PhoneticEncoder so we can mix and match them (i.e. they both implement the same interface called something like IPhoneticEncoder).

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3335-added-configuration-to-specify-max-code-lengths.yml (hunk 5 above)
Comment (TipzCM) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3335-added-configuration-to-specify-max-code-lengths.yml:6]: as discussed, the documentation update will happen in https://gitlab.com/simpatico.ai/cdr/-/issues/2780

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderEnum.java (hunk 1 above)
Comment (TipzCM) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderEnum.java]: done

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java (hunk 2 above)
Comment (TipzCM) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java]: updated

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3335-added-configuration-to-specify-max-code-lengths.yml (hunk 4 above)
Comment (TipzCM) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3335-added-configuration-to-specify-max-code-lengths.yml]: fixed

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchCustomSearchParamTest.java (hunk 6 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchCustomSearchParamTest.java:243]: updated

File: hapi-fhir-base/src/test/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapperTest.java (hunk 7 above)
Comment (TipzCM) [reply hapi-fhir-base/src/test/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapperTest.java]: added

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java (hunk 9 above)
Comment (TipzCM) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java]: Refactored into Utils class so IPhoneticEncoder is returned directly instead.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java (hunk 3 above)
Comment (TipzCM) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java]: Refactored out.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java (hunk 8 above)
Comment (TipzCM) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java]: Hmm. Is there more insight you can give me on this?

Does this comment still make sense now that we refactored out the "wrapper" class?

I think we'll need a switch statement either way to take enum value -> encoder.

Comment (TipzCM) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/context/phonetic/PhoneticEncoderWrapper.java]: updated

**Review by tadgh — APPROVED** [review]
